Add gate modifiers (#70)
QuTech-Delft/cQASM-spec: docs/language_specification/statements/instructions/unitary_instructions.md

> Unitary instructions include:
> 
> - [named gates](unitary_instructions.md#named-gates), and
> - compositions of [gate modifiers](unitary_instructions.md#gate-modifiers) acting on a named gate.

Reviewer: ```suggestion
Unitary instructions, commonly known as gates, define single-qubit or multi-qubit operations
that change the state of a qubit register in a reversible and deterministic fashion. They include:

- [named gates](unitary_instructions.md#named-gates), or
- compositions of [gate modifiers](unitary_instructions.md#gate-modifiers) acting on a named gate.
```

??? info "Grammar for unitary instructions"
    
    _unitary-instruction_:  
    &emsp; _gate_ _qubit-arguments_  

    _gate_:  
    &emsp; _inv-gate_  
    &emsp; _pow-gate_  
    &emsp; _ctrl-gate_  
    &emsp; _named-gate_  

    _inv-gate_:  
    &emsp; `inv.`gate
    
    _pow-gate_:  
    &emsp; `pow(`_parameter_`).`_gate_
    
    _ctrl-gate_:  
    &emsp; `ctrl.`_gate_

    _named-gate_:  
    &emsp; _identifier_ _parameters_~opt~  

    _parameters_:  
    &emsp; __`(`__ _parameter-list_ __`)`__

    _parameter-list_:  
    &emsp; _parameter_  
    &emsp; _parameter-list_ __`,`__ _parameter_

    _parameter_:  
    &emsp; _integer-literal_  
    &emsp; _floating-literal_

    _qubit-arguments_:  
    &emsp; _qubit-argument-list_

    _qubit-argument-list_:  
    &emsp; _qubit-argument_  
    &emsp; _qubit-argument-list_ __`,`__ _qubit-argument_

    _qubit-argument_:  
    &emsp; _qubit-variable_  
    &emsp; _qubit-index_

    _qubit-variable_:  
    &emsp; _identifier_

    _qubit-index_:  
    &emsp; _index_

## Named gates

Named gates define single-qubit or multi-qubit unitary operations

> that change the state of a qubit register in a deterministic fashion.

Reviewer: This description applies to unitary instructions in general and should then be moved above (see other suggestion)

| Y    | Pauli-Y                                  | **`Y q[0]`**            |
| Y90  | Rotation around the _y_-axis of $\pi/2$  | **`Y90 q[0]`**          |
| mY90 | Rotation around the _y_-axis of $-\pi/2$ | **`mY90 q[0]`**         |
| Z    | Pauli-Z                                  | **`Z q[0]`**            |
| S    | Phase gate                               | **`S q[0]`**            |
| Sdag | S dagger gate                            | **`Sdag q[0]`**         |
| T    | T                                        | **`T q[0]`**            |
| Tdag | T dagger gate                            | **`Tdag q[0]`**         |
| Rx   | Arbitrary rotation around _x_-axis       | **`Rx(pi) q[0]`**       |
| Ry   | Arbitrary rotation around _y_-axis       | **`Ry(pi) q[0]`**       |
| Rz   | Arbitrary rotation around _z_-axis       | **`Rz(pi) q[0]`**       |
| CNOT | Controlled-NOT gate                      | **`CNOT q[0], q[1]`**   |
| CZ   | Controlled-Z, Controlled-Phase           | **`CZ q[0], q[1]`**     |
| CR   | Controlled phase shift (arbitrary angle) | **`CR(pi) q[0], q[1]`** |
| CRk  | Controlled phase shift ($\pi/2^k$)       | **`CRk(2) q[0], q[1]`** |

## Gate modifiers

> A gate modifier is an operator _Q_ that takes a gate as input and returns a modified gate as output (_Q: gate → gate_).
> _I.e._, it transforms an arbitrary unitary $U$ into $U'$, based on the particular modifier that is applied.

Reviewer: ```suggestion
A gate modifier is an operator _Q_ that takes a gate as input and returns a modified gate as output (_Q: gate → gate_),
_i.e._, it transforms an arbitrary unitary $U$ into $U'$, based on the particular modifier that is applied.
```

We consider the following gate modifiers:

- The _inverse_ modifier **`inv`**, which modifies the gate $U$ into $U^\dagger$, _e.g._, **`inv.X q[0]`**.
- The _power_ modifier **`pow`**, which raises the gate $U$ to the specified power $a$, _i.e._, $U^a$,
  where $a$ is a float, _e.g._, **`pow(pi/2).X q[0]`**.
- The _control_ modifier **`ctrl`**, which changes the gate $U$ into the controlled-$U$ gate, _e.g._,
  **`ctrl.X q[0], q[1]`**, where the control qubit is prepended to the list of qubit operands. 

Since gate modifiers return gates, they can be applied in a nested manner,

> e.g., `inv.pow(2).X`, where a gate modifier `inv` is applied to the modified gate `pow(2).X`,
> which in turn resulted from applying the gate modifier `pow(2)` on the (named) gate `X`.

Reviewer: ```suggestion
_e.g._, **`inv.pow(2).X`**, where a gate modifier **`inv`** is applied to the modified gate **`pow(2).X`**,
which in turn resulted from applying the gate modifier **`pow(2)`** on the (named) gate **`X`**.
```

Note that a composition of gate modifiers acting on a named gate is itself a gate. 

??? info "Grammar for unitary instructions"
    
    _unitary-instruction_:  
    &emsp; _gate_ _qubit-arguments_  

    _gate_:  
    &emsp; _inv-gate_  
    &emsp; _pow-gate_  
    &emsp; _ctrl-gate_  
    &emsp; _named-gate_  

    _inv-gate_:  
    &emsp; `inv.`gate
    
    _pow-gate_:  
    &emsp; `pow(`_parameter_`).`_gate_
    
    _ctrl-gate_:  
    &emsp; `ctrl.`_gate_

    _named-gate_:  
    &emsp; _identifier_ _parameters_~opt~  

    _parameters_:  
    &emsp; __`(`__ _parameter-list_ __`)`__

    _parameter-list_:  
    &emsp; _parameter_  
    &emsp; _parameter-list_ __`,`__ _parameter_

    _parameter_:  
    &emsp; _integer-literal_  
    &emsp; _floating-literal_

    _qubit-arguments_:  
    &emsp; _qubit-argument-list_

    _qubit-argument-list_:  
    &emsp; _qubit-argument_  
    &emsp; _qubit-argument-list_ __`,`__ _qubit-argument_

    _qubit-argument_:  
    &emsp; _qubit-variable_  
    &emsp; _qubit-index_

    _qubit-variable_:  
    &emsp; _identifier_

    _qubit-index_:  
    &emsp; _index_

## Named gates

> A named gate is what is commonly referred to simply as a gate.
> We use the term named gate to distinguish them from compositions of gate modifiers acting on a named gate,
> which also results to be a gate.
> Named gates are, thus, the simplest form of unitary instructions.
> They can just be used on their own, or modified multiple times by gate modifiers. 

Reviewer: ```suggestion
Named gates comprise particular unitary operations that have been given their own unique label, _e.g._, 
the Hadamard gate **`H`** or the controlled-not gate **`CNOT`**.
In general, we simply refer to them as gates.
All recognized (named) gates are listed in the [Standard gate set](unitary_instructions.md#standard-gate-set).
Moreover, we use the term 'named gate' to distinguish them from unitary operations consisting of compositions of gate modifiers acting on named gates, _i.e._, modified gates.
Named gates are, thus, unmodified gates and can simply be used on their own or modified (multiple times) through gate modifiers. 
```

Author: Done.
- I only changed `acting on named gates` to `acting on a named gate` because the same text is used in 3 other places in the doc.
- And I've removed the line `The named gates that are supported by the cQASM language
are listed in the [standard gate set](unitary_instructions.md#standard-gate-set).` a bit further down in the section, because it's basically a duplicate of the line 65 above.

  where _a_ is a float, e.g., `pow(pi/2).X q[0]`.
- **control**: changes _U_ gate for controlled-U gate, e.g., `ctrl.X q[0], q[1]`, 
  where the control qubit is prepended to the list of qubit operands. 

Notice that, `inv` and `pow` map an _n_-qubit gate to an _n_-qubit gate,
and `ctrl` maps an _n_-qubit gate to a (_n_+1)-qubit gate.

Since gate modifiers return gates, they can be applied in a nested manner,
e.g., `inv.pow(2).X`, where a gate modifier `inv` is applied to the modified gate `pow(2).X`,
which in turn resulted from applying the gate modifier `pow(2)` on the (named) gate `X`.

The current version of the language only allows gate modifiers to act on single-qubit gates.

!!! info ""

    &emsp;**`inv.`**_gate qubit-argument_  
    &emsp;**`pow(`**_exponent_**`).`**_gate qubit-argument_  
    &emsp;**`ctrl.`**_gate qubit-arguments_

A few examples of gates are shown below.

!!! example ""

    ```
    // Inverse of X
    inv.X q

    // S gate implemented as a power of T
    pow(2).T q

    // CZ imlemented as a controlled-Z
    ctrl.Z q[0], q[1]

    // Composition of gate modifiers (modifiers are applied from right to left)
    ctrl.pow(1/2).inv.X q[0], q[1]

>     ```

Author: Added the beginning and end, and fixed some types all over.

Thanks for the short form for the super index.

BTW, what's the `$` for?

>     ```

Reviewer: The $ is for using MathType in markdown (allows you to write equations).
